{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "32ddd5c0-7727-4e8e-b9cb-697cdf3cd067",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:95% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML, Math, Markdown\n",
    "display(HTML(\"<style>.container { width:95% !important; }</style>\"))\n",
    "\n",
    "import sys\n",
    "import os\n",
    "coralme_dir = '/home/chris/zuniga/coralme/'\n",
    "sys.path.insert(0, coralme_dir)\n",
    "# code for enabling this notebook to work within cursor\n",
    "\n",
    "\n",
    "import importlib\n",
    "import coralme\n",
    "import coralme.solver.solver\n",
    "import coralme.builder.main\n",
    "import coralme.core.model\n",
    "import os\n",
    "import pandas as pd\n",
    "import json\n",
    "import cobra\n",
    "\n",
    "from coralme.builder.main import MEBuilder\n",
    "\n",
    "from Bio import Entrez, SeqIO\n",
    "Entrez.email = 'cdalldorf@sdsu.edu'\n",
    "\n",
    "# Function to download GenBank file\n",
    "def download_genbank(ref_id, output_dir, verbose = False):\n",
    "    try:\n",
    "        if verbose: print(f\"Fetching {ref_id}...\")\n",
    "        with Entrez.efetch(db=\"nucleotide\", id=ref_id, rettype=\"gb\", retmode=\"text\") as handle:\n",
    "            gb_record = handle.read()\n",
    "            output_file = os.path.join(output_dir, 'genome.gb')\n",
    "            with open(output_file, \"w\") as f:\n",
    "                f.write(gb_record)\n",
    "        if verbose: print(f\"Saved {ref_id} to {output_file}\")\n",
    "        return(1)\n",
    "    except Exception as e:\n",
    "        if verbos: print(f\"Error fetching {ref_id}: {e}\")\n",
    "        return(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "36cf6d99-78f0-40c0-874c-60deab54664c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP065866 skipped, model already exists\n",
      "CP041013 skipped, model already exists\n",
      "CP039749 skipped, model already exists\n",
      "CP065865 skipped, model already exists\n",
      "CP065867 skipped, model already exists\n",
      "LR590473 skipped, model already exists\n",
      "CP008749.1 skipped, model already exists\n",
      "AE004091.2 skipped, model already exists\n",
      "CP068238 skipped, model already exists\n",
      "CP014784 skipped, model already exists\n",
      "CP061848 skipped, model already exists\n",
      "CP061335 skipped, model already exists\n",
      "CP053697 skipped, model already exists\n",
      "CP070982 skipped, model already exists\n",
      "LS483372 skipped, model already exists\n",
      "CP012830 skipped, model already exists\n",
      "CP012831 skipped, model already exists\n",
      "CP008896 skipped, model already exists\n",
      "CP015225 skipped, model already exists\n",
      "CP060288 skipped, model already exists\n",
      "CP022562 skipped, model already exists\n",
      "CP050291 skipped, model already exists\n",
      "CP043179 skipped, model already exists\n",
      "AP024503 skipped, model already exists\n",
      "AP022324 skipped, model already exists\n",
      "CP022560 skipped, model already exists\n",
      "CP022561 skipped, model already exists\n",
      "CP000712 skipped, model already exists\n",
      "CP069317 skipped, model already exists\n",
      "CP038001 skipped, model already exists\n",
      "CP058975 skipped, model already exists\n",
      "CP063780 skipped, model already exists\n",
      "CP026386 skipped, model already exists\n",
      "CP024478 skipped, model already exists\n",
      "CP045359 skipped, model already exists\n",
      "CP045349 skipped, model already exists\n",
      "CP032419 skipped, model already exists\n",
      "CP045416 skipped, model already exists\n",
      "AP014655.1 skipped, model already exists\n",
      "CP020100 skipped, model already exists\n",
      "CP043320 skipped, model already exists\n",
      "CP073105 skipped, model already exists\n",
      "CP076683 skipped, model already exists\n"
     ]
    }
   ],
   "source": [
    "# go through each line, convert model, find files necessary, generate input jsons\n",
    "base_dir = os.path.join(coralme_dir, 'species_files', 'Pseudomonas_files')\n",
    "info_df = pd.read_excel(os.path.join(base_dir,'Strain_models_genome-information_predictedM9.xlsx'), index_col = 0)\n",
    "for index, row in info_df.iterrows():\n",
    "    if index == 'Reference': continue\n",
    "\n",
    "\n",
    "    # check if run\n",
    "    out_dir = base_dir+'/individual_species/'+index+'/'\n",
    "    model_name = row['organism'].replace(' ', '_')+'_'+row['strain'].replace(' ','_')\n",
    "    out_model = out_dir+'/outputs/MEModel-step3-'+model_name+'-TS.pkl'\n",
    "    if os.path.exists(out_model):\n",
    "        print(str(index)+' skipped, model already exists')\n",
    "        continue\n",
    "    \n",
    "    # make directory\n",
    "    if not os.path.exists(out_dir):\n",
    "        os.mkdir(out_dir)\n",
    "    if not os.path.exists(out_dir+'/inputs'):\n",
    "        os.mkdir(out_dir+'/inputs')\n",
    "    if not os.path.exists(out_dir+'/outputs'):\n",
    "        os.mkdir(out_dir+'/outputs')\n",
    "    \n",
    "    # pull genome.gb file from NCBI\n",
    "    if not os.path.exists(out_dir+'/inputs/genome.gb'):\n",
    "        download_genbank(index, out_dir+'/inputs')\n",
    "\n",
    "    # TODO - automatic download of Biolog files\n",
    "    # you tried this before without success, perhaps do manually\n",
    "    biolog_exists = False\n",
    "    \n",
    "    # let's create organism.json and input.json files\n",
    "    if biolog_exists:\n",
    "        input_json = {\n",
    "          \"m-model-path\": \"./inputs/model.json\",\n",
    "          \"genbank-path\": \"./inputs/genome.gb\",\n",
    "          \"biocyc.genes\": \"./inputs/genes.txt\",\n",
    "          \"biocyc.prots\": \"./inputs/proteins.txt\",\n",
    "          \"biocyc.TUs\": \"./inputs/TUs.txt\",\n",
    "          \"biocyc.RNAs\": \"./inputs/RNAs.txt\",\n",
    "          \"biocyc.seqs\": \"./inputs/sequences.fasta\",\n",
    "          \"df_gene_cplxs_mods_rxns\": \"./outputs/building_data/automated-org-with-refs.xlsx\",\n",
    "          \"out_directory\": \"./outputs\",\n",
    "          \"log_directory\": \"./outputs\",\n",
    "          \"run_bbh_blast\": True,\n",
    "          \"e_value_cutoff\": 1e-10,\n",
    "          \"dev_reference\": True,\n",
    "          \"include_pseudo_genes\": True,\n",
    "          \"locus_tag\" : \"locus_tag\"\n",
    "        }\n",
    "    else:\n",
    "        input_json = {\n",
    "          \"m-model-path\": \"./inputs/model.json\",\n",
    "          \"genbank-path\": \"./inputs/genome.gb\",\n",
    "          \"df_gene_cplxs_mods_rxns\": \"./outputs/building_data/automated-org-with-refs.xlsx\",\n",
    "          \"out_directory\": \"./outputs\",\n",
    "          \"log_directory\": \"./outputs\",\n",
    "          \"run_bbh_blast\": True,\n",
    "          \"e_value_cutoff\": 1e-10,\n",
    "          \"dev_reference\": True,\n",
    "          \"include_pseudo_genes\": True,\n",
    "          \"locus_tag\" : \"locus_tag\"\n",
    "        }\n",
    "    organism = {\n",
    "    \t\"ME-Model-ID\" : model_name,\n",
    "    \t\"growth_key\" : \"mu1\",\n",
    "    \t\"complex_cofactors\" : {},\n",
    "    \t\"lipoprotein_precursors\" : {},\n",
    "    \t\"lipid_modifications\" : [],\n",
    "    \t\"flux_of_biomass_constituents\" : {},\n",
    "    \t\"flux_of_lipid_constituents\" : {},\n",
    "    \t\"braun's_lipoprotein\" : [],\n",
    "    \t\"braun's_lpp_flux\" : -0.0,\n",
    "    \t\"braun's_murein_flux\" : -0.0,\n",
    "    \t\"gr_data_doublings_per_hour\" : [0, 0.6, 1.0, 1.5, 2.0, 2.5],\n",
    "        \"percent_dna_data\" : [0.0592, 0.0512, 0.0330, 0.0252, 0.0222, 0.0208],\n",
    "    \t\"defer_to_rxn_matrix\" : []\n",
    "    }\n",
    "    with open(out_dir+'organism.json', 'w') as f:\n",
    "        json.dump(organism, f, indent = 4)\n",
    "    with open(out_dir+'input.json', 'w') as f:\n",
    "        json.dump(input_json, f, indent = 4)\n",
    "\n",
    "    \n",
    "    # convert model file to json if necessary\n",
    "    matlab_file = base_dir+'/matlab-strain-models/'+index+'.mat'\n",
    "    json_file = out_dir+'/inputs/model.json'\n",
    "    if not os.path.exists(json_file) and os.path.exists(matlab_file):\n",
    "        model = cobra.io.load_matlab_model(matlab_file)\n",
    "        cobra.io.save_json_model(model, json_file)\n",
    "\n",
    "    # create the model\n",
    "    organism = out_dir+'organism.json'\n",
    "    inputs = out_dir+'input.json'\n",
    "    try:\n",
    "        builder = MEBuilder(*[organism, inputs])\n",
    "        builder.generate_files(overwrite=True)\n",
    "        builder.build_me_model(overwrite=True)\n",
    "    except Exception as e:\n",
    "        print(f\"Error building model: {e}\")\n",
    "        continue\n",
    "    \n",
    "    # gapfill the model\n",
    "    builder.troubleshoot(growth_key_and_value = { builder.me_model.mu : 0.001 })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb47c6df",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d4e4448",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
