<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>2. Description of Inputs &mdash; coralME 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="3. Arquitecture of coralME" href="coralMEArquitecture.html" />
    <link rel="prev" title="1. Getting started" href="GettingStarted.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            coralME
          </a>
              <div class="version">
                1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="GettingStarted.html">1. Getting started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">2. Description of Inputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Types-of-inputs">2.1. Types of inputs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Required">2.1.1. Required</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Optional">2.1.2. Optional</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Configuration">2.1.3. Configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Description">2.2. Description</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Genome-(genome.gb)">2.2.1. Genome (<strong>genome.gb</strong>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id1">2.2.1.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Requirements">2.2.1.2. Requirements</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#M-model-(m_model.json)">2.2.2. M-model (<strong>m_model.json</strong>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id2">2.2.2.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id3">2.2.2.2. Requirements</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Gene-dictionary-(genes.txt)-[optional]">2.2.3. Gene dictionary (<strong>genes.txt</strong>) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id4">2.2.3.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id5">2.2.3.2. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Notes">2.2.3.3. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Proteins-(proteins.txt)-[optional]">2.2.4. Proteins (<strong>proteins.txt</strong>) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id6">2.2.4.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id7">2.2.4.2. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id8">2.2.4.3. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#RNAs-(RNAs.txt)-[optional]">2.2.5. RNAs (<strong>RNAs.txt</strong>) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id9">2.2.5.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id10">2.2.5.2. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id11">2.2.5.3. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#TUs-(TUs.txt)-[optional]">2.2.6. TUs (<strong>TUs.txt</strong>) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id12">2.2.6.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id13">2.2.6.2. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id14">2.2.6.3. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Gene-sequences-(sequences.fasta)-[optional]">2.2.7. Gene sequences (<strong>sequences.fasta</strong>) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id15">2.2.7.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id16">2.2.7.2. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id17">2.2.7.3. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Configuration-of-paths-to-files-(inputs.json)">2.2.8. Configuration of paths to files (<strong>inputs.json</strong>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id18">2.2.8.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id19">2.2.8.2. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id20">2.2.8.3. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Configuration-of-parameters-(organism.json)">2.2.9. Configuration of parameters (<strong>organism.json</strong>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id21">2.2.9.1. Description</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id22">2.2.9.2. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id23">2.2.9.3. Notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="coralMEArquitecture.html">3. Arquitecture of coralME</a></li>
<li class="toctree-l1"><a class="reference internal" href="ManualCuration.html">4. How to manually curate a ME-model using coralME</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">5. Frequently Asked Questions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">coralME</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">2. </span>Description of Inputs</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/BasicInputs.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Description-of-Inputs">
<h1><span class="section-number">2. </span>Description of Inputs<a class="headerlink" href="#Description-of-Inputs" title="Permalink to this heading"></a></h1>
<p>coralME takes a total of 7 inputs, 2 required and 5 optional:</p>
<section id="Types-of-inputs">
<h2><span class="section-number">2.1. </span>Types of inputs<a class="headerlink" href="#Types-of-inputs" title="Permalink to this heading"></a></h2>
<section id="Required">
<h3><span class="section-number">2.1.1. </span>Required<a class="headerlink" href="#Required" title="Permalink to this heading"></a></h3>
<ol class="arabic simple">
<li><p><strong>Genome file</strong> (<strong>genome.gb</strong>)</p></li>
<li><p><strong>M-model</strong> (<strong>m_model.json</strong> or <strong>m_model.xml</strong>)</p></li>
</ol>
</section>
<section id="Optional">
<h3><span class="section-number">2.1.2. </span>Optional<a class="headerlink" href="#Optional" title="Permalink to this heading"></a></h3>
<p>Downloadable from an existing <strong>BioCyc</strong> database under <strong>Special SmartTables</strong>. If no optional files are provided, coralME complements them with <strong>genome.gb</strong></p>
<ol class="arabic simple" start="3">
<li><p><strong>Genes file</strong>, by default: <strong>genes.txt</strong></p></li>
<li><p><strong>RNAs file</strong>, by default: <strong>RNAs.txt</strong></p></li>
<li><p><strong>Proteins file</strong>, by default: <strong>proteins.txt</strong></p></li>
<li><p><strong>TUs file</strong>, by default: <strong>TUs.txt.</strong></p></li>
<li><p><strong>Sequences file</strong>, by default: <strong>sequences.fasta</strong></p></li>
</ol>
</section>
<section id="Configuration">
<h3><span class="section-number">2.1.3. </span>Configuration<a class="headerlink" href="#Configuration" title="Permalink to this heading"></a></h3>
<ol class="arabic simple" start="8">
<li><p><strong>Paths file</strong>, by default: <strong>inputs.json</strong></p></li>
<li><p><strong>Parameters file</strong>, by default: <strong>organism.json</strong></p></li>
</ol>
<p><img alt="Drawing" src="_images/inputs.png" /></p>
</section>
</section>
<section id="Description">
<h2><span class="section-number">2.2. </span>Description<a class="headerlink" href="#Description" title="Permalink to this heading"></a></h2>
<section id="Genome-(genome.gb)">
<h3><span class="section-number">2.2.1. </span>Genome (<strong>genome.gb</strong>)<a class="headerlink" href="#Genome-(genome.gb)" title="Permalink to this heading"></a></h3>
<section id="id1">
<h4><span class="section-number">2.2.1.1. </span>Description<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h4>
<p>The genome file contains provides coralME with:</p>
<ul class="simple">
<li><p>Gene annotations.</p></li>
<li><p>Gene sequences.</p></li>
</ul>
</section>
<section id="Requirements">
<h4><span class="section-number">2.2.1.2. </span>Requirements<a class="headerlink" href="#Requirements" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Locus tags (locus_tag or old_locus_tag) MUST be consistent with <strong>m_model.json</strong>. Make sure you download the same genome file that was used to reconstruct the M-model.</p></li>
<li><p>Has name <strong>genome.gb</strong>.</p></li>
<li><p>Genbank-compliant file. Must be read by BioPython correctly.</p></li>
<li><p>It must contain the entire genome sequence. Make sure to enable <strong>Customize View</strong>&gt;<strong>Show Sequence</strong> before downloading the genbank file from NCBI.</p></li>
</ol>
<p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/inputs/genome.gb">genome.gb</a><strong>and</strong><a class="reference external" href="./helper_files/pputida_tutorial/inputs/sequences.fasta">sequences.fasta</a></p>
</section>
</section>
<section id="M-model-(m_model.json)">
<h3><span class="section-number">2.2.2. </span>M-model (<strong>m_model.json</strong>)<a class="headerlink" href="#M-model-(m_model.json)" title="Permalink to this heading"></a></h3>
<section id="id2">
<h4><span class="section-number">2.2.2.1. </span>Description<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h4>
<p>The M-model provides coralME with the metabolic model components:</p>
<ul class="simple">
<li><p>Metabolic network (M-matrix)</p></li>
<li><p>Gene-protein-reaction associations</p></li>
<li><p>Environmental and internal constraints</p></li>
<li><p>Reaction subsystems</p></li>
<li><p>Biomass composition</p></li>
</ul>
</section>
<section id="id3">
<h4><span class="section-number">2.2.2.2. </span>Requirements<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Gene identifiers MUST be consistent with <strong>genome.gb</strong> locus_tag or old_locus_tag. Make sure you download the same genome file that was used to reconstruct the M-model.</p></li>
<li><p>Has name <strong>m_model.json</strong>.</p></li>
<li><p>COBRApy-compliant. Must be read by cobrapy-0.25.0.</p></li>
</ol>
<p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/inputs/m_model.json">m_model.json</a></p>
</section>
</section>
<section id="Gene-dictionary-(genes.txt)-[optional]">
<h3><span class="section-number">2.2.3. </span>Gene dictionary (<strong>genes.txt</strong>) [optional]<a class="headerlink" href="#Gene-dictionary-(genes.txt)-[optional]" title="Permalink to this heading"></a></h3>
<section id="id4">
<h4><span class="section-number">2.2.3.1. </span>Description<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h4>
<p><strong>genes.txt</strong> is a gene information table that can be downloaded from the <strong>All genes of organism SmartTable</strong> of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click <strong>Export</strong>&gt;<strong>to Spreadsheet File</strong>&gt;<strong>frame IDs</strong>. <strong>This file is optional and is meant to complement the information from</strong>genome.gb** in case the latter is missing genes.**</p>
<p><strong>genes.txt</strong> provides coralME with:</p>
<ul class="simple">
<li><p>Gene locus tags</p></li>
<li><p>Gene names</p></li>
<li><p>Gene annotations</p></li>
<li><p>Gene positions</p></li>
<li><p>Gene products (protein, tRNA, etc.)</p></li>
</ul>
</section>
<section id="id5">
<h4><span class="section-number">2.2.3.2. </span>Requirements<a class="headerlink" href="#id5" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Contains the index <strong>Gene Name</strong> and columns <strong>Accession-1</strong>, <strong>Left-End-Position</strong>, <strong>Right-End-Position</strong>, and <strong>Product</strong>.</p></li>
<li><p><strong>Accession-1</strong> MUST be consistent with the gene IDs in the GPRs of <strong>m_model.json</strong> and with the locus_tag (or old_locus_tag) in <strong>genome.gb</strong>.</p></li>
<li><p><strong>Gene Name</strong> is consistent with:</p>
<ul class="simple">
<li><p>Column <strong>Genes of polypeptide, complex, or RNA</strong> of <strong>proteins.txt</strong></p></li>
<li><p>Column <strong>Gene</strong> of <strong>RNAs.txt</strong></p></li>
<li><p>Column <strong>Genes of transcription unit</strong> of <strong>TUs.txt</strong></p></li>
<li><p>Gene identifiers in <strong>sequences.fasta</strong></p></li>
</ul>
</li>
<li><p><strong>Product</strong> is consistent with:</p>
<ul class="simple">
<li><p>Index of <strong>proteins.txt</strong></p></li>
<li><p>Index of <strong>RNAs.txt</strong></p></li>
</ul>
</li>
<li><p>Must be tab-separated</p></li>
</ol>
</section>
<section id="Notes">
<h4><span class="section-number">2.2.3.3. </span>Notes<a class="headerlink" href="#Notes" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>Requirements 3, 4 and 5</strong> regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
<li><p><strong>Left-End-Position</strong> and <strong>Right-End-Position</strong> do not need to be consistent with the positions <strong>genome.gb</strong>. coralME will keep the positions in <strong>genome.gb</strong> over the specified ones in <strong>genes.txt</strong></p></li>
</ul>
<p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/inputs/genes.txt">genes.txt</a></p>
</section>
</section>
<section id="Proteins-(proteins.txt)-[optional]">
<h3><span class="section-number">2.2.4. </span>Proteins (<strong>proteins.txt</strong>) [optional]<a class="headerlink" href="#Proteins-(proteins.txt)-[optional]" title="Permalink to this heading"></a></h3>
<section id="id6">
<h4><span class="section-number">2.2.4.1. </span>Description<a class="headerlink" href="#id6" title="Permalink to this heading"></a></h4>
<p><strong>proteins.txt</strong> is a protein complex information table that can be downloaded from the <strong>All proteins of organism SmartTable</strong> of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click <strong>Export</strong>&gt;<strong>to Spreadsheet File</strong>&gt;<strong>frame IDs</strong>. <strong>This file is optional and is meant to complement the information from</strong>genome.gb<strong>.</strong></p>
<p><strong>proteins.txt</strong> provides coralME with: * Protein complex compositions</p>
</section>
<section id="id7">
<h4><span class="section-number">2.2.4.2. </span>Requirements<a class="headerlink" href="#id7" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Contains the index <strong>(Proteins Complexes)</strong> and columns <strong>Common-Name</strong>, <strong>Genes of polypeptide, complex, or RNA</strong>, and <strong>Locations</strong>.</p></li>
<li><p><strong>(Proteins Complexes)</strong> is consistent with:</p>
<ul class="simple">
<li><p>Column <strong>Product</strong> of <strong>genes.txt</strong></p></li>
</ul>
</li>
<li><p><strong>Genes of polypeptide, complex, or RNA</strong> is consistent with:</p>
<ul class="simple">
<li><p>Index <strong>Gene Name</strong> of <strong>genes.txt</strong></p></li>
</ul>
</li>
<li><p>Must be tab-separated</p></li>
</ol>
</section>
<section id="id8">
<h4><span class="section-number">2.2.4.3. </span>Notes<a class="headerlink" href="#id8" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>Requirements 2, 3 and 4</strong> regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
</ul>
<p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/inputs/proteins.txt">proteins.txt</a></p>
</section>
</section>
<section id="RNAs-(RNAs.txt)-[optional]">
<h3><span class="section-number">2.2.5. </span>RNAs (<strong>RNAs.txt</strong>) [optional]<a class="headerlink" href="#RNAs-(RNAs.txt)-[optional]" title="Permalink to this heading"></a></h3>
<section id="id9">
<h4><span class="section-number">2.2.5.1. </span>Description<a class="headerlink" href="#id9" title="Permalink to this heading"></a></h4>
<p><strong>RNAs.txt</strong> is an RNA annotation table that can be downloaded from the <strong>All RNAs of organism SmartTable</strong> of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click <strong>Export</strong>&gt;<strong>to Spreadsheet File</strong>&gt;<strong>frame IDs</strong>. <strong>This file is optional and is meant to complement the information from</strong>genome.gb<strong>.</strong></p>
<p><strong>RNAs.txt</strong> provides coralME with:</p>
<ul class="simple">
<li><p>Genes annotated as RNA products (e.g. tRNA, rRNA, etc.)</p></li>
<li><p>RNA gene annotations (e.g. amino acids - tRNA associations)</p></li>
</ul>
</section>
<section id="id10">
<h4><span class="section-number">2.2.5.2. </span>Requirements<a class="headerlink" href="#id10" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Contains the index <strong>(All-tRNAs Misc-RNAs rRNAs)</strong> and columns <strong>Common-Name</strong>, and <strong>Gene</strong></p></li>
<li><p><strong>(All-tRNAs Misc-RNAs rRNAs)</strong> is consistent with:</p>
<ul class="simple">
<li><p>Column <strong>Product</strong> of <strong>genes.txt</strong></p></li>
</ul>
</li>
<li><p><strong>Gene</strong> is consistent with:</p>
<ul class="simple">
<li><p>Index <strong>Gene Name</strong> of <strong>genes.txt</strong></p></li>
</ul>
</li>
<li><p>Must be tab-separated</p></li>
</ol>
</section>
<section id="id11">
<h4><span class="section-number">2.2.5.3. </span>Notes<a class="headerlink" href="#id11" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>Requirements 2, 3 and 4</strong> regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
</ul>
<p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/inputs/RNAs.txt">RNAs.txt</a></p>
</section>
</section>
<section id="TUs-(TUs.txt)-[optional]">
<h3><span class="section-number">2.2.6. </span>TUs (<strong>TUs.txt</strong>) [optional]<a class="headerlink" href="#TUs-(TUs.txt)-[optional]" title="Permalink to this heading"></a></h3>
<section id="id12">
<h4><span class="section-number">2.2.6.1. </span>Description<a class="headerlink" href="#id12" title="Permalink to this heading"></a></h4>
<p><strong>TUs.txt</strong> is a transcription unit annotation table that can be downloaded from the <strong>All TUs of organism SmartTable</strong> of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click <strong>Export</strong>&gt;<strong>to Spreadsheet File</strong>&gt;<strong>frame IDs</strong>. <strong>This file is optional and is meant to complement the information from</strong>genome.gb<strong>.</strong></p>
<p><strong>TUs.txt</strong> provides coralME with:</p>
<ul class="simple">
<li><p>Co-transcribed genes (operons).</p></li>
<li><p>Direction of transcription.</p></li>
<li><p>TU IDs.</p></li>
</ul>
</section>
<section id="id13">
<h4><span class="section-number">2.2.6.2. </span>Requirements<a class="headerlink" href="#id13" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Contains the index <strong>Transcription-Units</strong> and columns <strong>Genes of transcription unit</strong>, and <strong>Direction</strong></p></li>
<li><p><strong>Genes of transcription unit</strong> is consistent with:</p>
<ul class="simple">
<li><p>Index <strong>Gene Name</strong> of <strong>genes.txt</strong></p></li>
</ul>
</li>
<li><p>Must be tab-separated</p></li>
</ol>
</section>
<section id="id14">
<h4><span class="section-number">2.2.6.3. </span>Notes<a class="headerlink" href="#id14" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>Requirements 2 and 3</strong> regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
</ul>
<p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/inputs/TUs.txt">TUs.txt</a></p>
</section>
</section>
<section id="Gene-sequences-(sequences.fasta)-[optional]">
<h3><span class="section-number">2.2.7. </span>Gene sequences (<strong>sequences.fasta</strong>) [optional]<a class="headerlink" href="#Gene-sequences-(sequences.fasta)-[optional]" title="Permalink to this heading"></a></h3>
<section id="id15">
<h4><span class="section-number">2.2.7.1. </span>Description<a class="headerlink" href="#id15" title="Permalink to this heading"></a></h4>
<p><strong>sequences.fasta</strong> is a nucleotide FASTA file that can be downloaded from the <strong>All genes of organism SmartTable</strong> of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click <strong>Export</strong>&gt;<strong>FASTA</strong>&gt;<strong>Find sequences</strong>. <strong>This file is optional and is meant to complement the information from</strong>genome.gb** in case the latter is missing genes.**</p>
<p><strong>sequences.fasta</strong> provides coralME with:</p>
<ul class="simple">
<li><p>Gene sequences</p></li>
</ul>
</section>
<section id="id16">
<h4><span class="section-number">2.2.7.2. </span>Requirements<a class="headerlink" href="#id16" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Gene identifiers are consistent with:</p>
<ul class="simple">
<li><p>Index <strong>Gene Name</strong> of <strong>genes.txt</strong></p></li>
</ul>
</li>
<li><p>Must be tab-separated</p></li>
</ol>
</section>
<section id="id17">
<h4><span class="section-number">2.2.7.3. </span>Notes<a class="headerlink" href="#id17" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>Requirements 1, 2 and 3</strong> regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
</ul>
<p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/inputs/sequences.fasta">sequences.fasta</a></p>
</section>
</section>
<section id="Configuration-of-paths-to-files-(inputs.json)">
<h3><span class="section-number">2.2.8. </span>Configuration of paths to files (<strong>inputs.json</strong>)<a class="headerlink" href="#Configuration-of-paths-to-files-(inputs.json)" title="Permalink to this heading"></a></h3>
<section id="id18">
<h4><span class="section-number">2.2.8.1. </span>Description<a class="headerlink" href="#id18" title="Permalink to this heading"></a></h4>
<p><strong>inputs.json</strong> is a JSON file containing paths to input files for coralME.</p>
<p><strong>inputs.json</strong> provides coralME with:</p>
<ul class="simple">
<li><p>Paths to input files</p></li>
</ul>
</section>
<section id="id19">
<h4><span class="section-number">2.2.8.2. </span>Requirements<a class="headerlink" href="#id19" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Must be JSON-compliant</p></li>
<li><p>Must contain paths to required files (M-model and Genome).</p></li>
<li><p>All defined files must exist.</p></li>
</ol>
</section>
<section id="id20">
<h4><span class="section-number">2.2.8.3. </span>Notes<a class="headerlink" href="#id20" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/input.json">input.json</a></p></li>
</ul>
</section>
</section>
<section id="Configuration-of-parameters-(organism.json)">
<h3><span class="section-number">2.2.9. </span>Configuration of parameters (<strong>organism.json</strong>)<a class="headerlink" href="#Configuration-of-parameters-(organism.json)" title="Permalink to this heading"></a></h3>
<section id="id21">
<h4><span class="section-number">2.2.9.1. </span>Description<a class="headerlink" href="#id21" title="Permalink to this heading"></a></h4>
<p><strong>organism.json</strong> is a JSON file containing paths to input files for coralME.</p>
<p><strong>organism.json</strong> provides coralME with:</p>
<ul class="simple">
<li><p>ME-modeling parameters</p></li>
</ul>
</section>
<section id="id22">
<h4><span class="section-number">2.2.9.2. </span>Requirements<a class="headerlink" href="#id22" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Must be JSON-compliant</p></li>
<li><p>Must contain the standard fields.</p></li>
</ol>
</section>
<section id="id23">
<h4><span class="section-number">2.2.9.3. </span>Notes<a class="headerlink" href="#id23" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p><strong>See an example of</strong><a class="reference external" href="./helper_files/pputida_tutorial/organism.json">organism.json</a></p></li>
</ul>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="GettingStarted.html" class="btn btn-neutral float-left" title="1. Getting started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="coralMEArquitecture.html" class="btn btn-neutral float-right" title="3. Arquitecture of coralME" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Juan D. Tibocha-Bonilla, Rodrigo Santibanez-Palominos.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>