<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>1. Getting started &mdash; coralME 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2. Description of Inputs" href="BasicInputs.html" />
    <link rel="prev" title="Welcome to coralME documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            coralME
          </a>
              <div class="version">
                1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">1. Getting started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Download-files-from-BioCyc">1.1. Download files from BioCyc</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Go-to-Tools&gt;Special-SmartTables">1.1.1. Go to Tools&gt;Special SmartTables</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Download-genes.txt-and-sequences.fasta">1.1.2. Download genes.txt and sequences.fasta</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Download-proteins.txt,-RNAs.txt-and-TUs.txt">1.1.3. Download proteins.txt, RNAs.txt and TUs.txt</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Download-proteins.txt">1.1.3.1. Download proteins.txt</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Download-RNAs.txt">1.1.3.2. Download RNAs.txt</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Download-TUs.txt">1.1.3.3. Download TUs.txt</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Initialize-the-folder-for-your-organism">1.2. Initialize the folder for your organism</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Define-inputs-in-input.json.">1.2.1. Define inputs in <strong>input.json</strong>.</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Define-parameters-in-organism.json.">1.2.2. Define parameters in <strong>organism.json</strong>.</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Reconstruct-with-coralME">1.3. Reconstruct with coralME</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-packages">1.3.1. Import packages</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Define-organism-and-inputs">1.3.2. Define organism and inputs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Create-builder">1.3.3. Create builder</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#coralme.builder.main.MEBuilder"><code class="docutils literal notranslate"><span class="pre">MEBuilder</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Generate-files">1.3.4. Generate files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Build-ME-model">1.3.5. Build ME-model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Troubleshoot-ME-model">1.3.6. Troubleshoot ME-model</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Curate-manually">1.4. Curate manually</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BasicInputs.html">2. Description of Inputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="coralMEArquitecture.html">3. Arquitecture of coralME</a></li>
<li class="toctree-l1"><a class="reference internal" href="ManualCuration.html">4. How to manually curate a ME-model using coralME</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">5. Frequently Asked Questions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">coralME</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">1. </span>Getting started</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/GettingStarted.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Getting-started">
<h1><span class="section-number">1. </span>Getting started<a class="headerlink" href="#Getting-started" title="Permalink to this heading"></a></h1>
<p>For details on inputs go to <a class="reference internal" href="BasicInputs.html"><span class="doc">Description of Inputs</span></a>.</p>
<p>For information about coralME architecture go to <a class="reference internal" href="coralMEArquitecture.html"><span class="doc">Arquitecture of coralME</span></a>.</p>
<section id="Download-files-from-BioCyc">
<h2><span class="section-number">1.1. </span>Download files from BioCyc<a class="headerlink" href="#Download-files-from-BioCyc" title="Permalink to this heading"></a></h2>
<p>BioCyc files are optional but useful, you should download them after having your gene id consistent M-model and genbank files.</p>
<p>The quickest way to do this is to copy one of the genes from the genbank file into the BioCyc search bar. Your organism should appear in the list if it is available in BioCyc.</p>
<p>To download:</p>
<section id="Go-to-Tools&gt;Special-SmartTables">
<h3><span class="section-number">1.1.1. </span>Go to Tools&gt;Special SmartTables<a class="headerlink" href="#Go-to-Tools>Special-SmartTables" title="Permalink to this heading"></a></h3>
<p><img alt="Drawing" src="_images/1.png" /></p>
<p>From here you can download the 5 optional BioCyc files for your organism.</p>
<p><img alt="Drawing" src="_images/2.png" /></p>
</section>
<section id="Download-genes.txt-and-sequences.fasta">
<h3><span class="section-number">1.1.2. </span>Download genes.txt and sequences.fasta<a class="headerlink" href="#Download-genes.txt-and-sequences.fasta" title="Permalink to this heading"></a></h3>
<p><img alt="Drawing" src="_images/3.png" /></p>
</section>
<section id="Download-proteins.txt,-RNAs.txt-and-TUs.txt">
<h3><span class="section-number">1.1.3. </span>Download proteins.txt, RNAs.txt and TUs.txt<a class="headerlink" href="#Download-proteins.txt,-RNAs.txt-and-TUs.txt" title="Permalink to this heading"></a></h3>
<p>The same process of genes.txt applies to proteins.txt, RNAs.txt and TUs.txt.</p>
<p>Some columns must be added manually using BioCyc’s dropdown lists <strong>ADD PROPERTY COLUMN</strong> and <strong>ADD TRANSFORM COLUMN</strong> within the SmartTable editing webpage.</p>
<p><img alt="Drawing" src="_images/4.png" /></p>
<section id="Download-proteins.txt">
<h4><span class="section-number">1.1.3.1. </span>Download proteins.txt<a class="headerlink" href="#Download-proteins.txt" title="Permalink to this heading"></a></h4>
<p>The index (Proteins Complexes) is in the SmartTable by default, but you need to add the columns Common-Name, Genes of polypeptide, complex, or RNA, and Locations.</p>
<ul class="simple">
<li><p>Common-Name is available in the dropdown list <strong>ADD PROPERTY COLUMN</strong></p></li>
<li><p>Genes of polypeptide, complex, or RNA is available in the dropdown list <strong>ADD TRANSFORM COLUMN</strong></p></li>
<li><p>Locations is available in the dropdown list <strong>ADD PROPERTY COLUMN</strong>.</p></li>
</ul>
</section>
<section id="Download-RNAs.txt">
<h4><span class="section-number">1.1.3.2. </span>Download RNAs.txt<a class="headerlink" href="#Download-RNAs.txt" title="Permalink to this heading"></a></h4>
<p>The index (All-tRNAs Misc-RNAs rRNAs) is in the SmartTable by default, but you need to add the columns Common-Name, and Gene.</p>
<ul class="simple">
<li><p>Common-Name is available in the dropdown list <strong>ADD PROPERTY COLUMN</strong></p></li>
<li><p>Gene is available in the dropdown list <strong>ADD PROPERTY COLUMN</strong>.</p></li>
</ul>
</section>
<section id="Download-TUs.txt">
<h4><span class="section-number">1.1.3.3. </span>Download TUs.txt<a class="headerlink" href="#Download-TUs.txt" title="Permalink to this heading"></a></h4>
<p>The index Transcription-Units is in the SmartTable by default, but you need to add the columns Genes of transcription unit, and Direction.</p>
<ul class="simple">
<li><p>Genes of transcription unit is available in the dropdown list <strong>ADD TRANSFORM COLUMN</strong></p></li>
<li><p>Direction is available in the dropdown list <strong>ADD PROPERTY COLUMN</strong>.</p></li>
</ul>
</section>
</section>
</section>
<section id="Initialize-the-folder-for-your-organism">
<h2><span class="section-number">1.2. </span>Initialize the folder for your organism<a class="headerlink" href="#Initialize-the-folder-for-your-organism" title="Permalink to this heading"></a></h2>
<p>Copy your files to create your initial folder</p>
<p><img alt="Drawing" src="_images/folder.png" /></p>
<section id="Define-inputs-in-input.json.">
<h3><span class="section-number">1.2.1. </span>Define inputs in <strong>input.json</strong>.<a class="headerlink" href="#Define-inputs-in-input.json." title="Permalink to this heading"></a></h3>
<p>See an example of <a class="reference external" href="./helper_files/tutorial/input.json">input.json</a></p>
</section>
<section id="Define-parameters-in-organism.json.">
<h3><span class="section-number">1.2.2. </span>Define parameters in <strong>organism.json</strong>.<a class="headerlink" href="#Define-parameters-in-organism.json." title="Permalink to this heading"></a></h3>
<p>See an example of <a class="reference external" href="./helper_files/tutorial/organism.json">organism.json</a></p>
<div class="admonition note">
<p><strong>Note</strong>: You do not need to modify these parameters right away. But once you are at the curation stage you will have to ensure these parameters are applicable to your organism.</p>
</div>
</section>
</section>
<section id="Reconstruct-with-coralME">
<h2><span class="section-number">1.3. </span>Reconstruct with coralME<a class="headerlink" href="#Reconstruct-with-coralME" title="Permalink to this heading"></a></h2>
<p>Here we show an example to reconstruct a dME-model of <em>B. subtilis</em></p>
<section id="Import-packages">
<h3><span class="section-number">1.3.1. </span>Import packages<a class="headerlink" href="#Import-packages" title="Permalink to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">coralme.builder.main</span> <span class="kn">import</span> <span class="n">MEBuilder</span>
</pre></div>
</div>
</div>
</section>
<section id="Define-organism-and-inputs">
<h3><span class="section-number">1.3.2. </span>Define organism and inputs<a class="headerlink" href="#Define-organism-and-inputs" title="Permalink to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">org</span> <span class="o">=</span> <span class="s2">&quot;./helper_files/tutorial/&quot;</span>
</pre></div>
</div>
</div>
<p>Load configuration files</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">org</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">organism</span> <span class="o">=</span> <span class="s1">&#39;./organism.json&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">org</span><span class="p">)</span>
<span class="n">inputs</span> <span class="o">=</span> <span class="s1">&#39;./input.json&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">org</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Create-builder">
<h3><span class="section-number">1.3.3. </span>Create builder<a class="headerlink" href="#Create-builder" title="Permalink to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.main.MEBuilder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.main.</span></span><span class="sig-name descname"><span class="pre">MEBuilder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/main.html#MEBuilder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.main.MEBuilder" title="Permalink to this definition"></a></dt>
<dd><p>MEBuilder class to coordinate the reconstruction of ME-models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>*args</strong> – Positional arguments are passed as paths to JSON files that
update the configuration of the parent class.</p></li>
<li><p><strong>**kwargs</strong> – Further keyword arguments are passed on as dictionaries
to update the configuration of the parent class.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="coralme.builder.main.MEBuilder.generate_files">
<span class="sig-name descname"><span class="pre">generate_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/main.html#MEBuilder.generate_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.main.MEBuilder.generate_files" title="Permalink to this definition"></a></dt>
<dd><p>Performs the Synchronize and Complement steps of the reconstruction.</p>
<p>This function will read the Organism and the Reference. It will
synchronize the input files, complement them, and finally build
the OSM for the Organism.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>overwrite</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a>) – If True, overwrite the OSM using the defined path in the configuration.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coralme.builder.main.MEBuilder.get_homology">
<span class="sig-name descname"><span class="pre">get_homology</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">evalue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/main.html#MEBuilder.get_homology"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.main.MEBuilder.get_homology" title="Permalink to this definition"></a></dt>
<dd><p>Calculates homology between Organism and Reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>evalue</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a><em>, </em><em>default 1e-10</em>) – Sets the E-value cutoff for calling protein
homologs using BLAST.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coralme.builder.main.MEBuilder.get_trna_to_codon">
<span class="sig-name descname"><span class="pre">get_trna_to_codon</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/main.html#MEBuilder.get_trna_to_codon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.main.MEBuilder.get_trna_to_codon" title="Permalink to this definition"></a></dt>
<dd><p>Gets tRNA to codon association from the Genome.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coralme.builder.main.MEBuilder.prepare_model">
<span class="sig-name descname"><span class="pre">prepare_model</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/main.html#MEBuilder.prepare_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.main.MEBuilder.prepare_model" title="Permalink to this definition"></a></dt>
<dd><p>Performs initial preparation of the M-model.</p>
<p>This function will fix some known issues that M-models can</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>overwrite</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a>) – If True, overwrite the OSM using the defined path in the configuration.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="coralme.builder.main.MEBuilder.troubleshoot">
<span class="sig-name descname"><span class="pre">troubleshoot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">growth_key_and_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">platform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">solver</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gurobi'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/main.html#MEBuilder.troubleshoot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.main.MEBuilder.troubleshoot" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>growth_key_and_value:</dt><dd><p>dictionary of Sympy.Symbol and value to replace</p>
</dd>
<dt>skip:</dt><dd><p>set of ME-components to not evaluate during gapfilling</p>
</dd>
<dt>platform:</dt><dd><p>‘win32’ to use gurobi (default) or cplex as solver</p>
</dd>
<dt>solver:</dt><dd><p>‘gurobi’ (default, if platform is ‘win32’) or ‘cplex’</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">builder</span> <span class="o">=</span> <span class="n">MEBuilder</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">organism</span><span class="p">,</span> <span class="n">inputs</span><span class="p">])</span>
</pre></div>
</div>
</div>
</section>
<section id="Generate-files">
<h3><span class="section-number">1.3.4. </span>Generate files<a class="headerlink" href="#Generate-files" title="Permalink to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">builder</span><span class="o">.</span><span class="n">generate_files</span><span class="p">(</span><span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Build-ME-model">
<h3><span class="section-number">1.3.5. </span>Build ME-model<a class="headerlink" href="#Build-ME-model" title="Permalink to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">builder</span><span class="o">.</span><span class="n">build_me_model</span><span class="p">(</span><span class="n">overwrite</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Troubleshoot-ME-model">
<h3><span class="section-number">1.3.6. </span>Troubleshoot ME-model<a class="headerlink" href="#Troubleshoot-ME-model" title="Permalink to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">builder</span><span class="o">.</span><span class="n">troubleshoot</span><span class="p">(</span><span class="n">growth_key_and_value</span> <span class="o">=</span> <span class="p">{</span> <span class="n">builder</span><span class="o">.</span><span class="n">me_model</span><span class="o">.</span><span class="n">mu</span> <span class="p">:</span> <span class="mf">0.001</span> <span class="p">})</span>
</pre></div>
</div>
</div>
<div class="admonition note">
<p><strong>Note</strong>: We set 0.001 as a standard value for feasibility checking, but feel free to modify it! Sometimes too high a value could put a significant strain on the model and give too many gaps to start with. Too low a value might not show you all the gaps needed.</p>
</div>
</section>
</section>
<section id="Curate-manually">
<h2><span class="section-number">1.4. </span>Curate manually<a class="headerlink" href="#Curate-manually" title="Permalink to this heading"></a></h2>
<p>For details on manual curation go to <a class="reference internal" href="ManualCuration.html"><span class="doc">How to manually curate a ME-model using coralME</span></a>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to coralME documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="BasicInputs.html" class="btn btn-neutral float-right" title="2. Description of Inputs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Juan D. Tibocha-Bonilla, Rodrigo Santibanez-Palominos.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>