<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>4. How to manually curate a ME-model using coralME &mdash; coralME 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="_static/thebelab.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="5. Frequently Asked Questions" href="FAQ.html" />
    <link rel="prev" title="3. Arquitecture of coralME" href="coralMEArquitecture.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            coralME
          </a>
              <div class="version">
                1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="GettingStarted.html">1. Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="BasicInputs.html">2. Description of Inputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="coralMEArquitecture.html">3. Arquitecture of coralME</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">4. How to manually curate a ME-model using coralME</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Manual-input-files">4.1. Manual input files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.TerminationSubreactions"><code class="docutils literal notranslate"><span class="pre">TerminationSubreactions</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.PeptideReleaseFactors"><code class="docutils literal notranslate"><span class="pre">PeptideReleaseFactors</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.RNADegradosome"><code class="docutils literal notranslate"><span class="pre">RNADegradosome</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.SpecialtRNASubreactions"><code class="docutils literal notranslate"><span class="pre">SpecialtRNASubreactions</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.LipoproteinPrecursors"><code class="docutils literal notranslate"><span class="pre">LipoproteinPrecursors</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.SpecialModifications"><code class="docutils literal notranslate"><span class="pre">SpecialModifications</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.ExcisionMachinery"><code class="docutils literal notranslate"><span class="pre">ExcisionMachinery</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.OrphanSpontReactions"><code class="docutils literal notranslate"><span class="pre">OrphanSpontReactions</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.EnzymeReactionAssociation"><code class="docutils literal notranslate"><span class="pre">EnzymeReactionAssociation</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.ProteinLocation"><code class="docutils literal notranslate"><span class="pre">ProteinLocation</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.TranslocationPathways"><code class="docutils literal notranslate"><span class="pre">TranslocationPathways</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.RNAModificationMachinery"><code class="docutils literal notranslate"><span class="pre">RNAModificationMachinery</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.RibosomeStoich"><code class="docutils literal notranslate"><span class="pre">RibosomeStoich</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.RhoIndependent"><code class="docutils literal notranslate"><span class="pre">RhoIndependent</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.Sigmas"><code class="docutils literal notranslate"><span class="pre">Sigmas</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.CleavedMethionine"><code class="docutils literal notranslate"><span class="pre">CleavedMethionine</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.FoldingDict"><code class="docutils literal notranslate"><span class="pre">FoldingDict</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.TranslocationMultipliers"><code class="docutils literal notranslate"><span class="pre">TranslocationMultipliers</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.SubreactionMatrix"><code class="docutils literal notranslate"><span class="pre">SubreactionMatrix</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.MEMetabolites"><code class="docutils literal notranslate"><span class="pre">MEMetabolites</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.SubsystemClassification"><code class="docutils literal notranslate"><span class="pre">SubsystemClassification</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.ReactionMatrix"><code class="docutils literal notranslate"><span class="pre">ReactionMatrix</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.AminoacidtRNASynthetase"><code class="docutils literal notranslate"><span class="pre">AminoacidtRNASynthetase</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.RNAModificationTargets"><code class="docutils literal notranslate"><span class="pre">RNAModificationTargets</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.TranscriptionSubreactions"><code class="docutils literal notranslate"><span class="pre">TranscriptionSubreactions</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.GenericDict"><code class="docutils literal notranslate"><span class="pre">GenericDict</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.RibosomeSubreactions"><code class="docutils literal notranslate"><span class="pre">RibosomeSubreactions</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#coralme.builder.curation.ReactionCorrections"><code class="docutils literal notranslate"><span class="pre">ReactionCorrections</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#How-to-curate?">4.2. How to curate?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">5. Frequently Asked Questions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">coralME</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">4. </span>How to manually curate a ME-model using coralME</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/ManualCuration.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="How-to-manually-curate-a-ME-model-using-coralME">
<h1><span class="section-number">4. </span>How to manually curate a ME-model using coralME<a class="headerlink" href="#How-to-manually-curate-a-ME-model-using-coralME" title="Permalink to this heading"></a></h1>
<section id="Manual-input-files">
<h2><span class="section-number">4.1. </span>Manual input files<a class="headerlink" href="#Manual-input-files" title="Permalink to this heading"></a></h2>
<p>After you run coralME for the first time the following files are generated in building_data/&lt;/code&gt;. Most of them are also automatically filled by the algorithm and saved in building_data/reference_files/&lt;/code&gt;. These _reference <a href="#id1"><span class="problematic" id="id2">files_</span></a> are meant to guide manual curation as they contain all information mapped by coralME formatted as manual input files.</p>
<p>__coralME does not overwrite any file in building_data/&lt;/code&gt;, but it will always overwrite files in building_data/reference_files/&lt;/code&gt;__</p>
<ul class="simple">
<li><dl class="simple">
<dt>termination_subreactions.txt</dt><dd><p>Input here will define translation termination subreactions and their
machinery.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.TerminationSubreactions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">TerminationSubreactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'termination_subreactions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'termination_subreactions.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Translation</span> <span class="pre">termination</span> <span class="pre">subreactions'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#TerminationSubreactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.TerminationSubreactions" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define translation termination subreactions.</p>
<p>This class creates the property “termination_subreactions” from
the manual inputs in termination_subreactions.txt in an
instance of Organism.</p>
<p>Input here will define translation termination subreactions and their
machinery.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>termination_subreactions.txt :</dt><dd><p>subreaction     enzymes
PrfA_mono_mediated_termination  PrfA_mono
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>peptide_release_factors.txt</dt><dd><p>Input here will define peptide release factors.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.PeptideReleaseFactors">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">PeptideReleaseFactors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'peptide_release_factors'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'peptide_release_factors.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Peptide</span> <span class="pre">release</span> <span class="pre">factors'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#PeptideReleaseFactors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.PeptideReleaseFactors" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define peptide release factors.</p>
<p>This class creates the property “peptide_release_factors” from
the manual inputs in peptide_release_factors.txt in an
instance of Organism.</p>
<p>Input here will define peptide release factors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>peptide_release_factors.txt :</dt><dd><p>release_factor  enzyme
UAA     generic_RF
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>rna_degradosome.txt</dt><dd><p>Input here will define the composition of the RNA degradosome.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.RNADegradosome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">RNADegradosome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rna_degradosome'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rna_degradosome.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RNA</span> <span class="pre">degradosome</span> <span class="pre">composition'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#RNADegradosome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.RNADegradosome" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to add RNA degradosome composition.</p>
<p>This class creates the property “rna_degradosome” from
the manual inputs in rna_degradosome.txt in an
instance of Organism.</p>
<p>Input here will define the composition of the RNA degradosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>rna_degradosome.txt :</dt><dd><p>enzymes
Eno_dim_mod_mg2(4)
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>special_trna_subreactions.txt</dt><dd><p>Input here will define special tRNA subreactions, such as
tRNA-Sec (selenocysteine) synthesis from tRNA-Ser.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.SpecialtRNASubreactions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">SpecialtRNASubreactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'special_trna_subreactions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'special_trna_subreactions.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Special</span> <span class="pre">tRNA</span> <span class="pre">subreactions'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#SpecialtRNASubreactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.SpecialtRNASubreactions" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define special tRNA subreactions.</p>
<p>This class creates the property “special_trna_subreactions” from
the manual inputs in special_trna_subreactions.txt in an
instance of Organism.</p>
<p>Input here will define special tRNA subreactions, such as
tRNA-Sec (selenocysteine) synthesis from tRNA-Ser.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>special_trna_subreactions.txt :</dt><dd><p>subreaction     enzymes
PrfA_mono_mediated_termination  PrfA_mono
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>lipoprotein_precursors.txt</dt><dd><p>Input here will add lipoprotein precursors.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.LipoproteinPrecursors">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">LipoproteinPrecursors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lipoprotein_precursors'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lipoprotein_precursors.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Lipoprotein</span> <span class="pre">precursors'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#LipoproteinPrecursors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.LipoproteinPrecursors" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to add lipoprotein precursors.</p>
<p>This class creates the property “lipoprotein_precursors” from
the manual inputs in lipoprotein_precursors.txt in an
instance of Organism.</p>
<p>Input here will add lipoprotein precursors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>lipoprotein_precursors.txt :</dt><dd><p>name,gene
AcrA,b0463
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>special_modifications.txt</dt><dd><p>Input here will define machinery for special modifications. These
modifications are a set of pre-defined modifications that are used
in ME-models.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.SpecialModifications">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">SpecialModifications</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'special_modifications'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'special_modifications.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Special</span> <span class="pre">protein</span> <span class="pre">modifications'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#SpecialModifications"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.SpecialModifications" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define machinery for special modifications.</p>
<p>This class creates the property “special_modifications” from
the manual inputs in special_modifications.txt in an
instance of Organism.</p>
<p>Input here will define machinery for special modifications. These
modifications are a set of pre-defined modifications that are used
in ME-models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>special_modifications.txt :</dt><dd><p>modification    enzymes stoich
fes_transfer    CPLX0-7617,IscA_tetra,CPLX0-7824
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>excision_machinery.txt</dt><dd><p>Input here will define machinery for excision.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.ExcisionMachinery">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">ExcisionMachinery</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'excision_machinery'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'excision_machinery.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Excision</span> <span class="pre">machinery'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#ExcisionMachinery"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.ExcisionMachinery" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define machinery for excision.</p>
<p>This class creates the property “excision_machinery” from
the manual inputs in excision_machinery.txt in an
instance of Organism.</p>
<p>Input here will define machinery for excision.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>excision_machinery.txt :</dt><dd><p>mechanism       enzymes
rRNA_containing RNase_E_tetra_mod_zn2(2), …
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>orphan_and_spont_reactions.txt</dt><dd><p>Input here will mark reactions as orphan or spontaneous. Orphan
reactions will be associated with CPLX_dummy, and spontaneous ones
will not require enzymes for flux.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.OrphanSpontReactions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">OrphanSpontReactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'orphan_and_spont_reactions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'orphan_and_spont_reactions.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Orphan</span> <span class="pre">and</span> <span class="pre">spontaneous</span> <span class="pre">reactions'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#OrphanSpontReactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.OrphanSpontReactions" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to add reactions to the ME-model.</p>
<p>This class creates the property “orphan_and_spont_reactions” from
the manual inputs in orphan_and_spont_reactions.txt in an
instance of Organism.</p>
<p>Input here will mark reactions as orphan or spontaneous. Orphan
reactions will be associated with CPLX_dummy, and spontaneous ones
will not require enzymes for flux.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>orphan_and_spont_reactions.txt :</dt><dd><p>name    description     is_reversible   is_spontaneous  subsystems
CODH_Fe_loading Loading of Fe   false   true
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>enzyme_reaction_association.txt</dt><dd><p>Input here will create the association between enzymes and
reactions in the ME-model.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.EnzymeReactionAssociation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">EnzymeReactionAssociation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'enz_rxn_assoc_df'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'enzyme_reaction_association.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Enzyme-reaction</span> <span class="pre">associations'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#EnzymeReactionAssociation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.EnzymeReactionAssociation" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to specify enzyme-reaction associations.</p>
<p>This class creates the property “enz_rxn_assoc_df” from
the manual inputs in enzyme_reaction_association.txt in an
instance of Organism.</p>
<p>Input here will create the association between enzymes and
reactions in the ME-model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>enzyme_reaction_association.txt :</dt><dd><p>Reaction        Complexes
ADNt2pp NUPG-MONOMER OR NUPC-MONOMER
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>peptide_compartment_and_pathways.txt</dt><dd><p>Input here will modify protein locations, and translocation
pathways in the ME-model.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.ProteinLocation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">ProteinLocation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'protein_location'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'peptide_compartment_and_pathways.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Protein</span> <span class="pre">location'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#ProteinLocation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.ProteinLocation" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to add protein locations.</p>
<p>This class creates the property “protein_location” from
the manual inputs in peptide_compartment_and_pathways.txt in an
instance of Organism.</p>
<p>Input here will modify protein locations, and translocation
pathways in the ME-model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>peptide_compartment_and_pathways.txt :</dt><dd><p>Complex Complex_compartment     Protein Protein_compartment     translocase_pathway
BSU02690-MONOMER        Plasma_Membrane BSU02690()      Plasma_Membrane s
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>translocation_pathways.txt</dt><dd><p>Input here will define translocation pathways and their
machinery.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.TranslocationPathways">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">TranslocationPathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'translocation_pathways'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'translocation_pathways.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Translocation</span> <span class="pre">pathways'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#TranslocationPathways"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.TranslocationPathways" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define translocation pathways.</p>
<p>This class creates the property “translocation_pathways” from
the manual inputs in translocation_pathways.txt in an
instance of Organism.</p>
<p>Input here will define translocation pathways and their
machinery.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>translocation_pathways.txt :</dt><dd><p>pathway enzyme
sec     BSU27650-MONOMER
sec     BSU35300-MONOMER
sec     secYEG
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>rna_modification.txt</dt><dd><p>Input here will define enzymes that perform RNA modifications
for either rRNA or tRNA in the ME-model.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.RNAModificationMachinery">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">RNAModificationMachinery</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rna_modification_df'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rna_modification.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RNA</span> <span class="pre">Modification</span> <span class="pre">machinery'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#RNAModificationMachinery"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.RNAModificationMachinery" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to add RNA modification machinery.</p>
<p>This class creates the property “rna_modification_df” from
the manual inputs in rna_modification.txt in an
instance of Organism.</p>
<p>Input here will define enzymes that perform RNA modifications
for either rRNA or tRNA in the ME-model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>rna_modification.txt :</dt><dd><p>modification    positions       type    enzymes source
D       16,17,20,20A,21 tRNA    DusB_mono
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>ribosomal_proteins.txt</dt><dd><p>Input here will define the composition of the ribosome.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.RibosomeStoich">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">RibosomeStoich</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ribosome_stoich'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ribosomal_proteins.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Ribosomal</span> <span class="pre">proteins'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#RibosomeStoich"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.RibosomeStoich" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define ribosome composition.</p>
<p>This class creates the property “ribosome_stoich” from
the manual inputs in ribosomal_proteins.txt in an
instance of Organism.</p>
<p>Input here will define the composition of the ribosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>ribosomal_proteins.txt :</dt><dd><p>subunits        proteins
30S     RpsD_mono,…
50S     generic_23s_rRNAs,generic_5s_rRNAs,RplA_mono,…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>rho_independent.txt</dt><dd><p>Input here will mark genes with rho independent transcription
termination.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.RhoIndependent">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">RhoIndependent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rho_independent'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rho_independent.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Genes</span> <span class="pre">with</span> <span class="pre">rho-independent</span> <span class="pre">termination'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#RhoIndependent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.RhoIndependent" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define genes with rho independent
termination.</p>
<p>This class creates the property “rho_independent” from
the manual inputs in rho_independent.txt in an
instance of Organism.</p>
<p>Input here will mark genes with rho independent transcription
termination.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>rho_independent.txt :</dt><dd><p>id
b0344
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>sigma_factors.txt</dt><dd><p>Input here will mark proteins for N-terminal methionine
cleavage in the ME-model.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.Sigmas">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">Sigmas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sigmas'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sigma_factors.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Sigma</span> <span class="pre">factors'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#Sigmas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.Sigmas" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to modify or add sigma factors.</p>
<p>This class creates the property “sigmas” from
the manual inputs in sigma_factors.txt in an
instance of Organism.</p>
<p>Input here will mark proteins for N-terminal methionine
cleavage in the ME-model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>sigma_factors.txt :</dt><dd><p>sigma,complex,genes,name
RpoH_mono,RNAP_32H,b3461,”RNA polymerase, sigma 32 (sigma H) factor”
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>cleaved_methionine.txt</dt><dd><p>Input here will mark proteins for N-terminal methionine
cleavage in the ME-model.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.CleavedMethionine">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">CleavedMethionine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cleaved_methionine'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cleaved_methionine.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Proteins</span> <span class="pre">with</span> <span class="pre">N-terminal</span> <span class="pre">methionine</span> <span class="pre">cleavage'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#CleavedMethionine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.CleavedMethionine" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to mark proteins that undergo
N-terminal methionine cleavage.</p>
<p>This class creates the property “cleaved_methionine” from
the manual inputs in cleaved_methionine.txt in an
instance of Organism.</p>
<p>Input here will mark proteins for N-terminal methionine
cleavage in the ME-model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>cleaved_methionine.txt :</dt><dd><p>cleaved_methionine_genes
b4154
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>folding_dict.txt</dt><dd><p>Input here will define folding pathways for proteins.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.FoldingDict">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">FoldingDict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'folding_dict'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'folding_dict.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Protein</span> <span class="pre">to</span> <span class="pre">folding</span> <span class="pre">machinery</span> <span class="pre">associations'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#FoldingDict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.FoldingDict" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define folding pathways for proteins.</p>
<p>This class creates the property “folding_dict” from
the manual inputs in folding_dict.txt in an
instance of Organism.</p>
<p>Input here will define folding pathways for proteins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>folding_dict.txt :</dt><dd><p>mechanism       enzymes
GroEL_dependent_folding b0014, …
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>translocation_multipliers.txt</dt><dd><p>Input here will modify how many pores are required for
the translocation of a protein.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.TranslocationMultipliers">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">TranslocationMultipliers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'translocation_multipliers'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'translocation_multipliers.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Translocation</span> <span class="pre">multipliers'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#TranslocationMultipliers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.TranslocationMultipliers" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define translocation multipliers.</p>
<p>This class creates the property “translocation_multipliers” from
the manual inputs in translocation_multipliers.txt in an
instance of Organism.</p>
<p>Input here will modify how many pores are required for
the translocation of a protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>translocation_multipliers.txt :</dt><dd><p>Gene,YidC_MONOMER,TatE_MONOMER,TatA_MONOMER
b1855,2.0,0.0,0.0
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>subreaction_matrix.txt</dt><dd><p>Input here will define subreactions in the ME-model.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.SubreactionMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">SubreactionMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'subreaction_matrix'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'subreaction_matrix.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Matrix</span> <span class="pre">of</span> <span class="pre">subreaction</span> <span class="pre">stoichiometries'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#SubreactionMatrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.SubreactionMatrix" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to add subreactions.</p>
<p>This class creates the property “subreaction_matrix” from
the manual inputs in subreaction_matrix.txt in an
instance of Organism.</p>
<p>Input here will define subreactions in the ME-model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>subreaction_matrix.txt :</dt><dd><p>Reaction        Metabolites     Stoichiometry
mod_acetyl_c    accoa_c -1.0
mod_acetyl_c    coa_c   +1.0
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>me_metabolites.txt</dt><dd><p>Input here will mark metabolites in the M-model for replacement
with their corrected E-matrix component.</p>
</dd>
</dl>
</li>
</ul>
<ul class="simple">
<li><dl class="simple">
<dt>elongation_subreactions.txt</dt><dd><p>Input here will define translation elongation subreactions and their
machinery.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.MEMetabolites">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">MEMetabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'me_mets'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'me_metabolites.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Metabolites</span> <span class="pre">to</span> <span class="pre">substitute</span> <span class="pre">from</span> <span class="pre">M-model'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#MEMetabolites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.MEMetabolites" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to replace metabolites in the M-model.</p>
<p>This class creates the property “me_mets” from
the manual inputs in me_metabolites.txt in an
instance of Organism.</p>
<p>Input here will mark metabolites in the M-model for replacement
with their corrected E-matrix component.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>me_metabolites.txt :</dt><dd><p>id      me_id   name    formula compartment     type
sufbcd_c        CPLX0-1341              SufBCD complex          REPLACE
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>subsystem_classification.txt</dt><dd><p>Input here will classify subsystems in umbrella classifications which
are then used to set a median Keff and correct it with the
complex SASA.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.SubsystemClassification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">SubsystemClassification</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'subsystem_classification'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'subsystem_classification.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Classification</span> <span class="pre">of</span> <span class="pre">subsystems'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#SubsystemClassification"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.SubsystemClassification" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to classify subsystems for Keff estimation.</p>
<p>This class creates the property “subsystem_classification” from
the manual inputs in subsystem_classification.txt in an
instance of Organism.</p>
<p>Input here will classify subsystems in umbrella classifications which
are then used to set a median Keff and correct it with the
complex SASA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>subsystem_classification.txt :
subsystem   central_CE      central_AFN     intermediate    secondary       other
S_Amino_acids_and_related_molecules 0       1       0       0       0
…</p>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>reaction_matrix.txt</dt><dd><p>Input here will define reactions directly in the
ME-model. Definitions here will be added to the ME-model
after processing the M-model into the ME-model.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.ReactionMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">ReactionMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reaction_matrix'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reaction_matrix.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Matrix</span> <span class="pre">of</span> <span class="pre">reaction</span> <span class="pre">stoichiometries'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#ReactionMatrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.ReactionMatrix" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to add reactions to the ME-model.</p>
<p>This class creates the property “reaction_matrix” from
the manual inputs in reaction_matrix.txt in an
instance of Organism.</p>
<p>Input here will define reactions directly in the
ME-model. Definitions here will be added to the ME-model
after processing the M-model into the ME-model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>reaction_matrix.txt :</dt><dd><p>Reaction        Metabolites     Stoichiometry
Cs_cyto_import  cs_p    -1.0
Cs_cyto_import  h_c     1.0
Cs_cyto_import  cs_c    1.0
Cs_cyto_import  h_p     -1.0
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>lipid_modifications.txt</dt><dd><p>Input here will define enzymes that perform lipid
modifications.</p>
</dd>
</dl>
</li>
</ul>
<ul class="simple">
<li><dl class="simple">
<dt>amino_acid_trna_synthetase.txt</dt><dd><p>Input here will define amino acid tRNA ligases.</p>
</dd>
</dl>
</li>
</ul>
<ul class="simple">
<li><dl class="simple">
<dt>initiation_subreactions.txt</dt><dd><p>Input here will define translation initiation subreactions and their
machinery.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.AminoacidtRNASynthetase">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">AminoacidtRNASynthetase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'amino_acid_trna_synthetase'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'amino_acid_trna_synthetase.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Amino</span> <span class="pre">acid</span> <span class="pre">to</span> <span class="pre">tRNA</span> <span class="pre">synthetase</span> <span class="pre">associations'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#AminoacidtRNASynthetase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.AminoacidtRNASynthetase" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define amino acid tRNA ligases.</p>
<p>This class creates the property “amino_acid_trna_synthetase” from
the manual inputs in amino_acid_trna_synthetase.txt in an
instance of Organism.</p>
<p>Input here will define amino acid tRNA ligases.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>amino_acid_trna_synthetase.txt :</dt><dd><p>amino_acid      enzyme
ala__L_c        Ala_RS_tetra_mod_zn2(4)
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>post_transcriptional_modification_of_RNA.txt</dt><dd><p>Input here will define RNA genes that undergo modifications.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.RNAModificationTargets">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">RNAModificationTargets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rna_modification_targets'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'post_transcriptional_modification_of_RNA.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RNA</span> <span class="pre">modification</span> <span class="pre">targets'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#RNAModificationTargets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.RNAModificationTargets" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to add RNA modification targets.</p>
<p>This class creates the property “rna_modification_targets” from
the manual inputs in post_transcriptional_modification_of_RNA.txt in an
instance of Organism.</p>
<p>Input here will define RNA genes that undergo modifications.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>post_transcriptional_modification_of_RNA.txt :</dt><dd><p>bnum    position        modification
b0202   20A     D
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>protein_corrections.txt</dt><dd><p>Input here will add, modify complexes in the ME-model,
as well as add, modify their modifications. You can
add a complex modification ID in the replace column,
which will remove that modified complex and replace
it with your manually added one.</p>
</dd>
</dl>
</li>
</ul>
<ul class="simple">
<li><dl class="simple">
<dt>reaction_median_keffs.txt</dt><dd><p>Input here will define median Keffs for estimation of Keffs
using the SASA method.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>transcription_subreactions.txt</dt><dd><p>Input here will define machinery for transcription subreactions. These
subreactions are a set of pre-defined subreactions that are used
in ME-models.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.TranscriptionSubreactions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">TranscriptionSubreactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'transcription_subreactions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'transcription_subreactions.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Transcription</span> <span class="pre">subreactions'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#TranscriptionSubreactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.TranscriptionSubreactions" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define transcription subreactions.</p>
<p>This class creates the property “transcription_subreactions” from
the manual inputs in transcription_subreactions.txt in an
instance of Organism.</p>
<p>Input here will define machinery for transcription subreactions. These
subreactions are a set of pre-defined subreactions that are used
in ME-models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>transcription_subreactions.txt :</dt><dd><p>mechanism       enzymes
Transcription_normal_rho_independent    Mfd_mono_mod_mg2(1),NusA_mono,NusG_mono,GreA_mono,GreB_mono,RpoZ_mono_mod_mg2(1)
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>generic_dict.txt</dt><dd><p>Input here will define generics.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.GenericDict">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">GenericDict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'generic_dict'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'generic_dict.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Dictionary</span> <span class="pre">of</span> <span class="pre">generic</span> <span class="pre">complexes'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#GenericDict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.GenericDict" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define generics.</p>
<p>This class creates the property “generic_dict” from
the manual inputs in generic_dict.txt in an
instance of Organism.</p>
<p>Input here will define generics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>generic_dict.txt :</dt><dd><p>generic_component       enzymes
generic_16Sm4Cm1402     RsmH_mono,RsmI_mono
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>ribosome_subreactions.txt</dt><dd><p>Input here will define enzymes that perform a ribosome
subreaction.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.RibosomeSubreactions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">RibosomeSubreactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ribosome_subreactions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ribosome_subreactions.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Ribosomal</span> <span class="pre">subreactions'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#RibosomeSubreactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.RibosomeSubreactions" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to define ribosome subreactions.</p>
<p>This class creates the property “ribosome_subreactions” from
the manual inputs in ribosome_subreactions.txt in an
instance of Organism.</p>
<p>Input here will define enzymes that perform a ribosome
subreaction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>ribosome_subreactions.txt :</dt><dd><p>subreaction     enzyme
gtp_bound_30S_assembly_factor_phase1    BSU16650-MONOMER
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>reaction_corrections.txt</dt><dd><p>Input here will modify reactions at the M-model stage
before ME-model building.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="coralme.builder.curation.ReactionCorrections">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coralme.builder.curation.</span></span><span class="sig-name descname"><span class="pre">ReactionCorrections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">org</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reaction_corrections'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reaction_corrections.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Reaction</span> <span class="pre">corrections'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/coralme/builder/curation.html#ReactionCorrections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#coralme.builder.curation.ReactionCorrections" title="Permalink to this definition"></a></dt>
<dd><p>Reads manual input to modify reactions in the M-model.</p>
<p>This class creates the property “reaction_corrections” from
the manual inputs in reaction_corrections.txt in an
instance of Organism.</p>
<p>Input here will modify reactions at the M-model stage
before ME-model building.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>org</strong> (<em>coralme.builder.organism.Organism</em>) – Organism object.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<dl class="simple">
<dt>reaction_corrections.txt :</dt><dd><p>reaction_id,name,gene_reaction_rule,reaction,notes
COBALT2tpp,cobalt transport in via permease (no H+),BSU24740,cobalt2_e –&gt; cobalt2_c,No notes
…</p>
</dd>
</dl>
</dd></dl>

<ul class="simple">
<li><dl class="simple">
<dt>TUs_from_biocyc.txt</dt><dd><p>Input here will modify transcriptional unit information.</p>
</dd>
</dl>
</li>
</ul>
</section>
<section id="How-to-curate?">
<h2><span class="section-number">4.2. </span>How to curate?<a class="headerlink" href="#How-to-curate?" title="Permalink to this heading"></a></h2>
<ol class="arabic simple">
<li><p>If you have not run coralME yet, go back to <a class="reference internal" href="GettingStarted.html"><span class="doc">GettingStarted.ipynb</span></a>.</p></li>
<li><p><strong>Copy</strong> all of the generated <em>reference files</em> in building_data/reference_files and replace accordingly in building_data/</p></li>
<li><p><strong>Go one by one</strong> through the files in building_data/ curating as needed! Important flags are risen in curation_notes.json to further guide you through curation.</p></li>
<li><p>Everytime you make a change, <strong>run the model through the troubleshooter!</strong> It will show you remaining gaps to look at, and the new curation notes might show new warnings.</p></li>
<li><p><strong>Keep iterating!</strong> You will have finished when no gaps are present, and all remaining warnings in curation notes are irrelevant.</p></li>
</ol>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="coralMEArquitecture.html" class="btn btn-neutral float-left" title="3. Arquitecture of coralME" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="FAQ.html" class="btn btn-neutral float-right" title="5. Frequently Asked Questions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Juan D. Tibocha-Bonilla, Rodrigo Santibanez-Palominos.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>